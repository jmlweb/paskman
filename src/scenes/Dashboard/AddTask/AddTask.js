// @flow
import React from 'react'
import Modal from '../../../components/Modal/ModalContainer';

const AddTask = () => {
  return <div>Hola</div>;
}

const AddTaskWithModal = ({ name }: { name: string }) => (
  <Modal name={name} title="Add new task">
    <AddTask />
  </Modal>
);

export default AddTaskWithModal;
