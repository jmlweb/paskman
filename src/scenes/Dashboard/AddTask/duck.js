import {
  createAction,
  handleActions,
} from 'redux-actions';
import Rx from 'rxjs';
import stateMock from '../../../stateMock';
import { tasksAdd } from '../../../data/tasks/duck';

/**
 * INITIAL STATE
 */
export const initialState = stateMock.scenes.dashboard.addTask;

/**
 * CONSTANTS
 */
export const ADD_TASK_NAME_CHANGE = 'ADD_TASK/NAME_CHANGE';
export const ADD_TASK_NAME_BLUR = 'ADD_TASK/NAME_BLUR';
export const ADD_TASK_DESCRIPTION_CHANGE = 'ADD_TASK/DESCRIPTION_CHANGE';
export const ADD_TASK_TOGGLE_TIMEMODE = 'ADD_TASK/TOGGLE_TIMEMODE';
export const ADD_TASK_TIME_CHANGE = 'ADD_TASK/TIME_CHANGE';
export const ADD_TASK_SET_DATA = 'ADD_TASK/SET_DATA';
export const ADD_TASK_SAVE = 'ADD_TASK/SAVE';
export const ADD_TASK_RESET = 'ADD_TASK/RESET';

/**
 * ACTIONS
 */

export const addTaskNameChange = createAction(ADD_TASK_NAME_CHANGE, value => ({
  value,
  isValid: value.length >= 3,
}));
export const addTaskNameBlur = createAction(ADD_TASK_NAME_BLUR, hasBlur => ({
  hasChanged: hasBlur,
}));
export const addTaskDescriptionChange = createAction(ADD_TASK_DESCRIPTION_CHANGE);
export const addTaskToggleTimeMode = createAction(ADD_TASK_TOGGLE_TIMEMODE);
export const addTaskTimeChange = createAction(ADD_TASK_TIME_CHANGE, value => +value);
export const addTaskSetData = createAction(ADD_TASK_SET_DATA);
export const addTaskSave = createAction(ADD_TASK_SAVE);
export const addTaskReset = createAction(ADD_TASK_RESET);

/**
 * SELECTORS
 */
export const timeToPomodorosSelector = state =>
  Math.round(state.scenes.dashboard.addTask.timeRequired / state.data.settings.target.working);

/**
 * EPICS
 */
export const addTaskNameChangeEpic = (action$, store) => {
  const getCurrentName = state => state.scenes.dashboard.addTask.name;
  return action$
    .filter(action => action.type === ADD_TASK_NAME_CHANGE || action.type === ADD_TASK_NAME_BLUR)
    .mergeMap(({ payload }) => Rx.Observable.of(addTaskSetData({
      name: { ...getCurrentName(store.getState()), ...payload },
    })));
};

export const addTaskDescriptionChangeEpic = action$ =>
  action$.ofType(ADD_TASK_DESCRIPTION_CHANGE)
    .mergeMap(({ payload }) =>
      Rx.Observable.of(addTaskSetData({
        description: payload,
      })));

export const addTaskToggleTimeModeEpic = (action$, store) => {
  return action$.ofType(ADD_TASK_TOGGLE_TIMEMODE)
    .mergeMap(() => {
      const addTaskState = store.getState().scenes.dashboard.addTask;
      const targetTime = store.getState().data.settings.target.working;
      return Rx.Observable.of(addTaskSetData({
        timeRequired: addTaskState.timeMode
          ? timeToPomodorosSelector(store.getState()) * targetTime
          : addTaskState.timeRequired,
        timeMode: !addTaskState.timeMode,
      }));
    });
};

export const addTaskTimeChangeEpic = action$ =>
  action$.ofType(ADD_TASK_TIME_CHANGE)
    .mergeMap(({ payload }) => Rx.Observable.of(addTaskSetData({
      timeRequired: payload,
    })));

export const addTaskSaveEpic = action$ =>
  action$.ofType(ADD_TASK_SAVE)
    .mergeMap((action) => {
      const { callback, ...newTask } = action.payload;
      return Rx.Observable.concat(
        Rx.Observable.of(addTaskSetData({
          isSaving: true,
        })),
        Rx.Observable.of(tasksAdd({
          type: 'pending',
          newTask,
        })).debounceTime(1000).delay(1000),
        Rx.Observable.of(addTaskSetData({
          isSaving: false,
        })),
        Rx.Observable.of(addTaskReset()),
        Rx.Observable.of(callback),
      );
    });

/**
 * REDUCER
 */

export default handleActions({
  [ADD_TASK_SET_DATA]: (state, { payload }) => ({ ...state, ...payload }),
  [ADD_TASK_RESET]: () => initialState,
}, initialState);
