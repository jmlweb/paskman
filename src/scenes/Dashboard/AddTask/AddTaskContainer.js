// @flow
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { modalToggle } from '../../../components/Modal/duck';
import AddTask from './AddTask';

class AddTaskContainer extends Component {
  constructor(props) {
    super(props);
  }
  render() {
    return <AddTask />;
  }
}

export function mapStateToProps(state: {
  components: {
    modal: mixed,
  }
}) {
  return { ...state.components.modal };
};

export const mapDispatchToProps = {
  modalToggle,
};

const AddTaskConnectedContainer = connect(
  mapStateToProps,
  mapDispatchToProps
)(AddTaskContainer);

export default AddTaskConnectedContainer;
