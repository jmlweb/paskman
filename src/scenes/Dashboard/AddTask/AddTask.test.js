import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import { AddTask } from './AddTask';

describe('Heading', () => {
  it('Render the component', () => {
    const wrapper = mount(<AddTask />);
    expect(wrapper.find(AddTask).length).toEqual(1);
  });
  it('Matchs snapshot', () => {
    const wrapper = mount(<AddTask />);
    const tree = renderer.create(wrapper).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
