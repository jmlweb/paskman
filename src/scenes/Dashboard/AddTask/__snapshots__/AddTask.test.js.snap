// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Heading Matchs snapshot 1`] = `
<form
  className="sc-bZQynM hSufOU"
>
  <fieldset
    className="sc-gzVnrw cJCIgP"
  >
    <div
      className="sc-htoDjs hpWqMX"
    >
      <label
        className="sc-dnqmqq jebcPH"
      >
        Name (3 chars min)
      </label>
      <input
        className="sc-iwsKbI WHWzN"
        id="name"
        minLength="3"
        placeholder="Try to be concise"
        required={true}
        type="text"
      />
    </div>
    <div
      className="sc-htoDjs hpWqMX"
    >
      <label
        className="sc-dnqmqq jebcPH"
      >
        Description
      </label>
      <textarea
        className="sc-gZMcBi jEvzYf"
        cols="80"
        id="description"
        placeholder="Describe the task"
        rows="4"
      />
    </div>
    <div
      className="sc-htoDjs hpWqMX"
    >
      <label
        className="sc-dnqmqq jebcPH"
      >
        Pomodoros required
      </label>
      <div
        className="sc-gqjmRU hpLqBK"
      >
        <label
          className="sc-VigVT dzzzHZ"
          selected={true}
        >
          <input
            className="sc-jTzLTM hbqTmN"
            id="id-25"
            onClick={undefined}
            type="radio"
            value={25}
          />
          <span>
            1
          </span>
        </label>
        <label
          className="sc-VigVT cnzhie"
          selected={false}
        >
          <input
            className="sc-jTzLTM hbqTmN"
            id="id-50"
            onClick={undefined}
            type="radio"
            value={50}
          />
          <span>
            2
          </span>
        </label>
        <label
          className="sc-VigVT cnzhie"
          selected={false}
        >
          <input
            className="sc-jTzLTM hbqTmN"
            id="id-75"
            onClick={undefined}
            type="radio"
            value={75}
          />
          <span>
            3
          </span>
        </label>
        <label
          className="sc-VigVT cnzhie"
          selected={false}
        >
          <input
            className="sc-jTzLTM hbqTmN"
            id="id-100"
            onClick={undefined}
            type="radio"
            value={100}
          />
          <span>
            4
          </span>
        </label>
        <label
          className="sc-VigVT cnzhie"
          selected={false}
        >
          <input
            className="sc-jTzLTM hbqTmN"
            id="id-125"
            onClick={undefined}
            type="radio"
            value={125}
          />
          <span>
            5
          </span>
        </label>
      </div>
      <div
        className="sc-jzJRlG bzLuwb"
      >
        <button
          className="sc-kpOJdX eBApnI"
          color="primary"
          size="sm"
        >
          Change to time
        </button>
      </div>
    </div>
  </fieldset>
  <button
    className="sc-dxgOiQ exCsrx"
    color="success"
    size="md"
    type="submit"
  >
    Save
  </button>
</form>
`;
